app-engine/java/testSrc/com/google/cloud/tools/intellij/appengine/java/cloud/AppEngineArtifactDeploymentSourceProviderTest.java

```diff
@@ -0,0 +1,183 @@
+
+  @TestModule private Module module1;
+  @TestModule private Module module2;
+
+  @TestService @Mock AppEngineProjectService projectService;
+
+  private AppEngineArtifactDeploymentSourceProvider artifactDeploymentSourceProvider;
+
+  @Before
+  public void setUp() {
+    artifactDeploymentSourceProvider = new AppEngineArtifactDeploymentSourceProvider();
+  }
+
+  @Test
+  public void getDeploymentSources_withOnlyStandardArtifacts_returnsAllSources() {
+    makeStandard(module1);
+    makeStandard(module2);
+
+    assertDeploymentSourcesPresent(AppEngineEnvironment.APP_ENGINE_STANDARD);
+  }
+
+  @Test
+  public void getDeploymentSources_withOnlyFlexibleArtifacts_returnsAllSources() {
+    makeFlexible(module1);
+    makeFlexible(module2);
+
+    assertDeploymentSourcesPresent(AppEngineEnvironment.APP_ENGINE_FLEX);
+  }
+
+  @Test
+  public void getDeploymentSources_withStandardAndFlexibleArtifacts_returnsAllSources() {
+    makeStandard(module1);
+    makeFlexible(module2);
+
+    List<DeploymentSource> deploymentSources =
+        artifactDeploymentSourceProvider.getDeploymentSources(testFixture.getProject());
+
+    assertThat(deploymentSources.size()).isEqualTo(2);
+
+    DeploymentSource deploymentSource1 = deploymentSources.get(0);
+    DeploymentSource deploymentSource2 = deploymentSources.get(1);
+
+    assertThat(deploymentSource1 instanceof AppEngineArtifactDeploymentSource).isTrue();
+    assertThat(deploymentSource2 instanceof AppEngineArtifactDeploymentSource).isTrue();
+    assertThat(((AppEngineArtifactDeploymentSource) deploymentSource1).getEnvironment())
+        .isEqualTo(AppEngineEnvironment.APP_ENGINE_STANDARD);
+    assertThat(((AppEngineArtifactDeploymentSource) deploymentSource2).getEnvironment())
+        .isEqualTo(AppEngineEnvironment.APP_ENGINE_FLEX);
+  }
+
+  private void assertDeploymentSourcesPresent(AppEngineEnvironment environment) {
+    List<DeploymentSource> deploymentSources =
+        artifactDeploymentSourceProvider.getDeploymentSources(testFixture.getProject());
+
+    assertThat(deploymentSources.size()).isEqualTo(2);
+
+    DeploymentSource deploymentSource1 = deploymentSources.get(0);
+    DeploymentSource deploymentSource2 = deploymentSources.get(1);
+
+    assertThat(deploymentSource1 instanceof AppEngineArtifactDeploymentSource).isTrue();
+    assertThat(deploymentSource2 instanceof AppEngineArtifactDeploymentSource).isTrue();
+    assertThat(((AppEngineArtifactDeploymentSource) deploymentSource1).getEnvironment())
+        .isEqualTo(environment);
+    assertThat(((AppEngineArtifactDeploymentSource) deploymentSource2).getEnvironment())
+        .isEqualTo(environment);
+  }
+
+  /**
+   * Adds the {@link AppEngineStandardFacet} to the module, associates the module with the standard
+   * environment, and creates a new artifact associated with the module output.
+   *
+   * @param module
+   */
+  private void makeStandard(Module module) {
```

Comment:
Could it make sense to add those as a general test utility methods to `ModuleTestUtils`? They look clean and straighforward, might be useful for other tests too.